File: 8.无重复字符的最长子串/code/8.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """
        无重复字符的最长子串 - 滑动窗口法
        
        时间复杂度：O(n)
        空间复杂度：O(min(m, n))
        
        算法步骤：
            1. 使用字典记录每个字符最后出现的位置
            2. 右指针遍历字符串
            3. 如果遇到重复字符，移动左指针到重复字符之后
            4. 更新最大长度
        
        Args:
            s: 输入字符串
            
        Returns:
            int：无重复字符的最长子串长度
        """
        if not s:
            return 0
        
        char_index = {}
        max_length = 0
        left = 0
        
        for right, char in enumerate(s):
            if char in char_index and char_index[char] >= left:
                left = char_index[char] + 1
            
            char_index[char] = right
            max_length = max(max_length, right - left + 1)
        
        return max_length

File: 8.无重复字符的最长子串/8.cpp
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;
class Solution {
public:
    /**
     * 无重复字符的最长子串 - 滑动窗口法
     * 
     * 时间复杂度：O(n)
     * 空间复杂度：O(min(m, n))
     */
    int lengthOfLongestSubstring(string s) {
        if (s.empty()) {
            return 0;
        }
        
        unordered_map<char, int> char_index;
        int max_length = 0;
        int left = 0;
        
        for (int right = 0; right < s.length(); right++) {
            char c = s[right];
            
            if (char_index.find(c) != char_index.end() && char_index[c] >= left) {
                left = char_index[c] + 1;
            }
            
            char_index[c] = right;
             max_length = max(max_length, right - left + 1);
         }
         
         return max_length;
     }
};